knimin/lib/data_access.py

```diff
@@ -3532,6 +3533,204 @@ def read_dna_plate(self, dna_plate_id):
+	    The sample volume in nanoliters for each well
+	plate_normalization_water : 2D np.array of float
+	    The water volume in nanoliters for each well
+
+	Raises
+	------
+	ValueError
+	    If the shotgun plate `shotgun_plate_id` does not exist
+	    If `plate_normalization_sample` dimensions doesn't match the plate
+	        type
+	    If `plate_normalization_water` dimensions doesn't match the plate
+	        type
+	    If the echo instrument does not exist
+
+	Returns
+	-------
+	int
+	    The created normalized_shotgun_plate_id
+	"""
+        with TRN:
+            # verify the plate exists
+            sql = """SELECT EXISTS(
+                        SELECT 1
+                        FROM pm.shotgun_plate
+                        WHERE shotgun_plate_id = %s)"""
+            TRN.add(sql, [shotgun_plate_id])
+            res = TRN.execute_fetchindex()[0]
+            if not res[0]:
+                raise ValueError("shotgun plate %s does not exist" %
+                                 shotgun_plate_id)
+
+            # obtain and verify the echo exists
+            sql = """SELECT echo_id
+                     FROM pm.echo
+                     WHERE echo.name = %s"""
+            TRN.add(sql, [echo])
+            res = TRN.execute_fetchindex()
+            if not res:
+                raise ValueError("echo machine %s does not exist" % echo)
+            else:
+                echo_id = dict(res[0])['echo_id']
+
+            # verify the shape of the plate
+            sql = """SELECT rows, cols
+                     FROM pm.shotgun_plate JOIN
+                          pm.plate_type USING(plate_type_id)
+                     WHERE shotgun_plate_id = %s"""
+            TRN.add(sql, [shotgun_plate_id])
+            res = dict(TRN.execute_fetchindex()[0])
+            n_rows = res['rows']
+            n_cols = res['cols']
+
+            if plate_normalization_sample.shape != (n_rows, n_cols):
+                obs = repr(plate_normalization_sample.shape)
+                exp = repr((n_rows, n_cols))
+                raise ValueError("plate_normalization_sample has an unexpected "
+                                 "shape %s; expected %s" % (obs, exp))
+
+            if plate_normalization_water.shape != (n_rows, n_cols):
+                obs = repr(plate_normalization_water.shape)
+                exp = repr((n_rows, n_cols))
+                raise ValueError("plate_normalization_water has an unexpected "
+                                 "shape %s; expected %s" % (obs, exp))
+
+            # create the instance of the normalized plate
+            sql = """INSERT INTO pm.shotgun_normalized_plate
+                        (shotgun_plate_id, created_on, email, echo_id)
+                     VALUES (%s, now(), %s, %s)
+                     RETURNING shotgun_normalized_plate_id"""
+
+            TRN.add(sql, [shotgun_plate_id, email, echo_id])
+            shotgun_normalized_plate_id = TRN.execute_fetchindex()[0][0]
```

Comment:
Minor, but you can use `TRN.execute_fetchlast()`. is equivalent to `TRN.execute_fetchindex()[0][0]`